src/main/java/com/google/firebase/database/connection/NettyWebSocketClient.java

```diff
@@ -0,0 +1,202 @@
+      eventHandler.onError(e);
+    }
+  }
+
+  @Override
+  public void close() {
+    checkState(channel != null, "channel not initialized");
+    try {
+      channel.close();
+    } finally {
+      group.shutdownGracefully();
+      executorService.shutdown();
+      // TODO(hkj): https://github.com/netty/netty/issues/7310
+    }
+  }
+
+  @Override
+  public void send(String msg) {
+    checkState(channel != null && channel.isActive(), "channel not connected for sending");
+    channel.writeAndFlush(new TextWebSocketFrame(msg));
+  }
+
+  /**
+   * Handles low-level IO events. These events fire on the firebase-websocket-worker thread. We
+   * notify the {@link WebsocketConnection} on all events, which then hands them off to the
+   * RunLoop for further processing.
+   */
+  private static class WebSocketClientHandler extends SimpleChannelInboundHandler<Object> {
+
+    private final WebsocketConnection.WSClientEventHandler delegate;
+    private final WebSocketClientHandshaker handshaker;
+
+    WebSocketClientHandler(
+        URI uri, String userAgent, WebsocketConnection.WSClientEventHandler delegate) {
+      this.delegate = checkNotNull(delegate, "delegate must not be null");
+      checkArgument(!Strings.isNullOrEmpty(userAgent), "user agent must not be null or empty");
+      this.handshaker = WebSocketClientHandshakerFactory.newHandshaker(
+          uri, WebSocketVersion.V13, null, true,
+          new DefaultHttpHeaders().add("User-Agent", userAgent));
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext context) {
+      // Do nothing
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext context) {
+      handshaker.handshake(context.channel());
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext context) {
+      delegate.onClose();
+    }
+
+    @Override
+    public void channelRead0(ChannelHandlerContext context, Object message) throws Exception {
+      Channel channel = context.channel();
+      if (!handshaker.isHandshakeComplete()) {
+        checkState(message instanceof FullHttpResponse);
+        try {
+          handshaker.finishHandshake(channel, (FullHttpResponse) message);
+          delegate.onOpen();
+        } catch (WebSocketHandshakeException e) {
+          delegate.onError(e);
+        }
+        return;
+      }
+
+      if (message instanceof FullHttpResponse) {
+        FullHttpResponse response = (FullHttpResponse) message;
+        String error = String.format("Unexpected FullHttpResponse (status: %s; content: %s)",
+            response.status().toString(), response.content().toString(CharsetUtil.UTF_8));
+        throw new IllegalStateException(error);
+      }
+
+      WebSocketFrame frame = (WebSocketFrame) message;
+      if (frame instanceof TextWebSocketFrame) {
+        delegate.onMessage(((TextWebSocketFrame) frame).text());
+      } else if (frame instanceof CloseWebSocketFrame) {
+        delegate.onClose();
+      }
```

Comment:
should we have an else case that throws just as a sanity check that we're covering all cases?

Reply:
Done